### PR TITLE
Guarantee no resource version

### PR DESCRIPTION
**Pull request by leonorfmartins**
**What is this feature?**

This does several things:

- adds asyncronicity to Mode 3
- adds a condition on apiserver to only accept requests with a populated UID
- simplifies the dual writer layer A LOT after https://github.com/grafana/grafana/pull/93714 where we delegate to each resource to provide the needed payload for the dual writer to work.
- adds a condition to reject requests on which the resource version is populated

**Which issue(s) does this PR fix?**:
Part of https://github.com/grafana/search-and-storage-team/issues/96

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

File: pkg/apiserver/rest/dualwriter_mode2.go [hunk 1]
```diff
@@ -48,39 +48,44 @@ func (d *DualWriterMode2) Mode() DualWriterMode {
...
+	if err != nil {
+		return createdFromLegacy, err
+	}
+	if accIn.GetUID() != "" {
+		return nil, fmt.Errorf("there is an UID and it should not: %v", accIn.GetUID())
 	}
 
 	startStorage := time.Now()
-	rsp, err := d.Storage.Create(ctx, created, createValidation, options)
+	createdFromStorage, err := d.Storage.Create(ctx, in, createValidation, options)
```
Comment (ryantxu) [inline pkg/apiserver/rest/dualwriter_mode2.go:74]: I am still skeptical that we are not using `createdFromLegacy` as the input to this second Create.

At the very least, we need to ensure that the names returned from both are the same.  and fail if not?

**Review by ryantxu — APPROVED** [review]
This is worlds better than where we were!   I'm still not sure about name creation, but if the tests pass so far this is a good progress

File: pkg/storage/unified/resource/server.go [hunk 2]
```diff
@@ -222,6 +222,14 @@ func (s *server) newEvent(ctx context.Context, user claims.AuthInfo, key *Resour
...
 	}
 
+	if obj.GetUID() == "" {
+		return nil, NewBadRequestError("object is missing UID")
+	}
+
+	if obj.GetResourceVersion() != "" {
+		return nil, NewBadRequestError("object must not include resourceVersion")
+	}
+
```
Comment (ryantxu) [inline pkg/storage/unified/resource/server.go:232]: 🎉

**Comment by leonorfmartins** [comment]
There's a work in progress PR to fix the name thing, but that's breaking some other tests (see https://github.com/grafana/grafana/pull/93852)

I can add that check for now and see what happens. I'm just trying to keep PRs small-(ish) without breaking too much stuff 😅

**Comment by leonorfmartins** [comment]
we can't merge this PR because update is somehow creating another entry in storage DB, although the UIDs match and a previous object is found within GuaranteedUpdate. We will need more investigation on this.

**Comment by leonorfmartins** [comment]
closing this in favor of https://github.com/grafana/grafana/pull/93852 since that PR also checks for the same name to exist.

**Pull request closed without merging**